File: packages/control-client/typescript/src/protocol.ts
import {
  CborEnvelopeCodec, ClientError, encodeEnvelope, encodeFrame, encodeRecord, readExactly,
  type ByteTransport, type EstablishContext, type Established, type InboundFrame, type Protocol,
  type SendMetadata, type RawFrame,
} from "@microsandbox/protocol-client";
import {
  CONTROL_GENERATION, CONTROL_PROTOCOL, DEFAULT_MAX_IN_FLIGHT, DEFAULT_REQUEST_TIMEOUT_MS,
  DEFAULT_SETUP_TIMEOUT_MS, MAX_HANDSHAKE_FRAME_SIZE, decodeControlError, decodeWelcome,
  validateHello, type ControlHello, type ControlWelcome,
} from "./records.js";

/** Setup metadata retains the original welcome envelope and all unknown fields. */
export type ControlReady = { welcome: ControlWelcome; frame: InboundFrame };

/** Explicit framed setup; discovery and JSON adaptation belong outside this engine. */
export class ControlProtocol implements Protocol<ControlReady> {
  async establish(transport: ByteTransport, context: EstablishContext): Promise<Established<ControlReady>> {
    const hello: ControlHello = {
      protocol: CONTROL_PROTOCOL, min_generation: CONTROL_GENERATION, max_generation: CONTROL_GENERATION,
      max_frame_size: context.limits.maxFrameSize,
      max_in_flight: Math.min(context.limits.maxInFlight, DEFAULT_MAX_IN_FLIGHT),
    };
    try { validateHello(hello); } catch { throw new ClientError("invalid_options"); }
    await transport.write(encodeFrame({ id: 0, flags: 0, body: encodeEnvelope({
      v: CONTROL_GENERATION, t: "control.hello", p: encodeRecord(hello),
    }) }));
    const prefix = await readExactly(transport, 4, context.signal);
    const length = new DataView(prefix.buffer, prefix.byteOffset, prefix.byteLength).getUint32(0);
    // Reject hostile opening lengths before allocating or waiting for the body.
    if (length < 5 || length > MAX_HANDSHAKE_FRAME_SIZE) throw new ClientError("invalid_data");
    const header = await readExactly(transport, 5, context.signal);
    const body = await readExactly(transport, length - 5, context.signal);
    const codec = new CborEnvelopeCodec();
    const { frame, welcome } = decodeOpening(codec, { id: new DataView(header.buffer).getUint32(0), flags: header[4]!, body }, hello);
    return {
      transport, codec, ids: { start: 1, endExclusive: 0x100000000 }, ready: { welcome, frame },
      limits: {
        ...context.limits, maxFrameSize: welcome.max_frame_size, maxInFlight: welcome.max_in_flight,
        incompleteFrameTimeoutMs: context.limits.incompleteFrameTimeoutMs ?? DEFAULT_SETUP_TIMEOUT_MS,
        requestTimeoutMs: context.limits.requestTimeoutMs ?? DEFAULT_REQUEST_TIMEOUT_MS,
      },
    };
  }
  prepare(ready: ControlReady, wireName: string): SendMetadata {
    if (wireName === "control.hello" || wireName === "control.welcome") throw new ClientError("unsupported_operation");
    return { generation: ready.welcome.generation, flags: 0 };
  }
}

function decodeOpening(codec: CborEnvelopeCodec, raw: RawFrame, hello: ControlHello): ControlReady {
  try {
    const frame = codec.decode(raw);
    if (frame.id !== 0 || frame.flags !== 1 || frame.protocolVersion !== CONTROL_GENERATION) throw new ClientError("invalid_data");
    if (frame.type === "control.error") {
      const refusal = decodeControlError(frame.payload);
      throw new ClientError(refusal.code === "unsupported_generation" ? "unsupported_operation" : "invalid_data");
    }
    if (frame.type !== "control.welcome") throw new ClientError("invalid_data");
    return { frame, welcome: decodeWelcome(frame.payload, hello) };
  } catch (error) {
    // Decoder failures describe peer data, not socket I/O. Keep diagnostics
    // sanitized and preserve an explicit unsupported-generation refusal.
    if (error instanceof ClientError) throw error;
    throw new ClientError("invalid_data");
  }
}

File: packages/control-client/rust/lib/protocol.rs
//! Framed control setup. This path never probes or falls back to JSON.

use std::sync::Arc;

use microsandbox_protocol::{
    codec::{self, RawFrame},
    control::{
        CONTROL_GENERATION, ControlError, ControlHello, ControlWelcome, DEFAULT_MAX_IN_FLIGHT,
        DEFAULT_REQUEST_TIMEOUT, DEFAULT_SETUP_TIMEOUT, MAX_HANDSHAKE_FRAME_SIZE,
    },
    wire::Envelope,
};
use microsandbox_protocol_client::{
    BoxFuture, BoxTransport, CborEnvelopeCodec, Client, ClientError, ClientResult, ConnectOptions,
    ErrorKind, Established, IdRange, Protocol, SendMetadata,
};
use tokio::io::AsyncReadExt;

//--------------------------------------------------------------------------------------------------
// Types
//--------------------------------------------------------------------------------------------------

/// Always-framed host control, including the full generic low-level surface.
pub type ControlClient = Client<ControlProtocol>;

/// Generation-one hello/welcome and operation metadata.
pub struct ControlProtocol;

/// Negotiated limits and the exact original welcome envelope.
#[derive(Debug, Clone)]
pub struct ControlReady {
    /// Peer-selected generation and application ceilings.
    pub welcome: ControlWelcome,
    /// Original frame, including unknown envelope fields.
    pub frame: RawFrame,
}

//--------------------------------------------------------------------------------------------------
// Trait Implementations
//--------------------------------------------------------------------------------------------------

impl Protocol for ControlProtocol {
    type Ready = ControlReady;

    fn establish(
        mut stream: BoxTransport,
        options: ConnectOptions,
    ) -> BoxFuture<'static, ClientResult<Established<Self::Ready>>> {
        Box::pin(async move {
            let hello = ControlHello {
                max_frame_size: options.limits.max_frame_size,
                max_in_flight: options
                    .limits
                    .max_in_flight
                    .min(DEFAULT_MAX_IN_FLIGHT as usize) as u32,
                ..Default::default()
            };
            hello
                .validate()
                .map_err(|_| ClientError::new(ErrorKind::InvalidOptions))?;
            let opening =
                Envelope::new(CONTROL_GENERATION, "control.hello", &hello)?.frame(0, 0)?;
            codec::write_raw_frame(&mut stream, &opening)
                .await
                .map_err(|error| match error {
                    microsandbox_protocol::ProtocolError::Io(error) => ClientError::from(error),
                    _ => ClientError::new(ErrorKind::InvalidData),
                })?;
            // Setup is already inside the engine's one total deadline. Bound
            // allocation from the opening prefix before reading any body.
            let length = stream.read_u32().await?;
            if !(5..=MAX_HANDSHAKE_FRAME_SIZE).contains(&length) {
                return Err(ClientError::new(ErrorKind::InvalidData));
            }
            let id = stream.read_u32().await?;
            let flags = stream.read_u8().await?;
            let mut body = vec![0; length as usize - 5];
            stream.read_exact(&mut body).await?;
            let frame = RawFrame { id, flags, body };
            let envelope = Envelope::decode(&frame.body)?;
            if frame.id != 0 || frame.flags != 1 || envelope.v != CONTROL_GENERATION {
                return Err(ClientError::new(ErrorKind::InvalidData));
            }
            if envelope.t == "control.error" {
                let refusal: ControlError = envelope.payload()?;
                let kind = if refusal.code == "unsupported_generation" {
                    ErrorKind::UnsupportedOperation
                } else {
                    ErrorKind::InvalidData
                };
                return Err(ClientError::new(kind));
            }
            if envelope.t != "control.welcome" {
                return Err(ClientError::new(ErrorKind::InvalidData));
            }
            let welcome: ControlWelcome = envelope.payload()?;
            welcome
                .validate_for(&hello)
                .map_err(|_| ClientError::new(ErrorKind::InvalidData))?;
            let mut limits = options.limits;
            limits.max_frame_size = welcome.max_frame_size;
            limits.max_in_flight = welcome.max_in_flight as usize;
            limits
                .incomplete_frame_timeout
                .get_or_insert(DEFAULT_SETUP_TIMEOUT);
            limits
                .request_timeout
                .get_or_insert(DEFAULT_REQUEST_TIMEOUT);
            Ok(Established {
                transport: stream,
                codec: Arc::new(CborEnvelopeCodec),
                ids: IdRange {
                    start: 1,
                    end_exclusive: 1u64 << 32,
                },
                ready: ControlReady { welcome, frame },
                limits,
            })
        })
    }

    fn prepare(ready: &Self::Ready, wire_name: &str) -> ClientResult<SendMetadata> {
        // Setup messages cannot be sent as ordinary named operations. Unknown
        // extensions remain possible; raw callers can inspect any wire shape.
        if matches!(wire_name, "control.hello" | "control.welcome") {
            return Err(ClientError::new(ErrorKind::UnsupportedOperation));
        }
        Ok(SendMetadata {
            generation: ready.welcome.generation,
            flags: 0,
        })
    }
}
